subiquity/server/contract_selection.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2022 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+""" Module that deals with contract token selection. """
+
+import asyncio
+import logging
+from typing import Any, List
+import time
+
+from subiquity.server.ubuntu_advantage import (
+    UAInterface,
+    APIUsageError,
+    )
+
+from subiquitycore.async_helpers import schedule_task
+
+
+log = logging.getLogger("subiquity.server.contract_selection")
+
+
+class UPCSExpiredError(Exception):
+    """ Exception to be raised when a contract selection expired. """
+
+
+class UnknownError(Exception):
+    """ Exception to be raised in case of unexpected error. """
+    def __init__(self, message: str = "", errors: List[Any] = []) -> None:
+        self.errors = errors
+        super().__init__(message)
+
+
+class ContractSelection:
+    """ Represents an already initiated contract selection. """
+    def __init__(
+            self,
+            client: UAInterface,
+            magic_token: str,
+            user_code: str,
+            validity_seconds: int) -> None:
+        """ Initialize the contract selection. """
+        self.client = client
+        self.magic_token = magic_token
+        self.user_code = user_code
+        self.validity_seconds = validity_seconds
+        self.task = asyncio.create_task(self._run_polling())
+
+    @classmethod
+    async def initiate(cls, client: UAInterface) \
+            -> "ContractSelection":
+        """ Initiate a contract selection and return a ContractSelection
+        request object. """
+        answer = await client.strategy.magic_initiate_v1()
+
+        if answer["result"] != "success":
+            raise UnknownError(errors=answer["errors"])
+
+        return cls(
+                client=client,
+                magic_token=answer["data"]["attributes"]["token"],
+                validity_seconds=answer["data"]["attributes"]["expires_in"],
+                user_code=answer["data"]["attributes"]["user_code"])
+
+    async def _run_polling(self) -> str:
+        """ Runs the polling and eventually return a contract token. """
+        # Wait an initlal 30 seconds before sending the first request, then
```

Comment:
minor typo

Reply:
Nice catch, I couldn't see it ever after you said it :+1: